### ISSUE TITLE
[v2] NO_AGENT error when attempting to intercept

### ISSUE THREAD

**Issue by foleymic** [telepresenceio/telepresence#1816]
Starting with version 2.3.0 of telepresence, I am getting the having the following error when attempting to intercept a workload - `Failed to establish intercept: intercept in error state NO_AGENT: No agent found for my-workload`

```
telepresence list -i 

    Intercept name: my-workload
    State         : error: NO_AGENT: No agent found for "my-workload"
    Workload kind : Deployment
    Destination   : 127.0.0.1:8080
    Intercepting  : using mechanism="http" with args=["--match=x-telepresence-intercept-id=mfoley"]
```

I see that the workload does indeed have the agent and it, along with my component are both up and running in the pod.  I only see this problem if I end up using the agent that supports preview urls (and http-match).  If I run the same intercept command without being logged in, it works.

This problem does not occur if I use versions of telepresence prior to 2.3.0 (e.g. v2.2.2).  

Also, I discovered that if my laptop has something running on the local port the intercept succeeds.  For example - 

```
# Note that the component I am intercepting is not NGINX, I just need something running locally on this port
docker run -d -p 8080:80 --name nginx nginx

telepresence login
telepresence intercept my-workload --port 8080


# OUTPUT:
intercepted
    Intercept name  : my-workload
    State           : ACTIVE
    Workload kind   : Deployment
    Destination     : 127.0.0.1:8080
    Intercepting    : HTTP requests that match all headers:
      'x-telepresence-intercept-id: 936e5047-a7de-4a07-b7a5-17cb7ec26811:my-workload'
    Preview URL     : https://some-url
    Layer 5 Hostname: xxxx
```

I feel like I must have something misconfigured or I must be doing something wrong, otherwise I would expect others to have reported this, but I am not able to figure this out.  I have seen this problem on all currently available versions of v2.3.x (v2.3.0, v2.3.1 and v2.3.2)

Any help would be very much appreciated.

**Comment by foleymic** [comment]
I forgot to mention that my k8s cluster is running in Google's GKE and I'm using google's kubectl (`gcloud components install kubectl`)

**Comment by thallgren** [comment]
@foleymic What exact version of telepresence are you using?

**Comment by foleymic** [comment]
Hi @thallgren.  I'm currently testing with v2.3.2 
![image](https://user-images.githubusercontent.com/6004328/123332245-42171700-d50e-11eb-862d-5ef3bb87f9e2.png)

**Comment by foleymic** [comment]
Here are the logs from the traffic-agent - 
```
time="2021-06-24 21:12:02.7052" level=info msg=" 1 Sending to intercept 9eed5daa-a45a-4692-bd4f-a49f8e92e740:xxxx when:" PID=1 THREAD=/sidecar
time="2021-06-24 21:12:02.7122" level=info msg="Accept got connection from 127.0.0.1:59038" PID=1 THREAD=/sidecar lis=":0"
time="2021-06-24 21:12:02.7497" level=error msg="tcp 127.0.0.1:59038 -> 127.0.0.1:8080: unhandled connection control message: tcp 127.0.0.1:59038 -> 127.0.0.1:8080, code CONNECT_REJECT, len 0" PID=1 THREAD=/sidecar lis=":0"
time="2021-06-24 21:12:20.8047" level=error msg="stream Recv: EOF" PID=1 THREAD=/sidecar
time="2021-06-24 21:12:20.8050" level=error msg="error establishing gRPC connection: context canceled" PID=1 THREAD=/sidecar
time="2021-06-24 21:12:20.8061" level=info msg="Error on accept: accept tcp [::]:46655: use of closed network connection" PID=1 THREAD=/sidecar
time="2021-06-24 21:12:20.8092" level=error msg="depart session: rpc error: code = Canceled desc = context canceled" PID=1 THREAD=/sidecar
time="2021-06-24 21:12:20.8095" level=error msg="talking to manager: rpc error: code = Canceled desc = context canceled" PID=1 THREAD=/sidecar
[2021-06-24 21:12:20.812][14][info][upstream] [source/common/upstream/cds_api_impl.cc:65] cds: add 2 cluster(s), remove 1 cluster(s)
[2021-06-24 21:12:20.814][14][info][upstream] [source/common/upstream/cds_api_impl.cc:80] cds: add/update cluster 'tel-proxy-46655'
time="2021-06-24 21:12:20.8168" level=info msg="Connected to Manager v2.3.2" PID=1 THREAD=/sidecar
[2021-06-24 21:12:20.821][14][info][upstream] [source/common/upstream/cds_api_impl.cc:65] cds: add 1 cluster(s), remove 2 cluster(s)
[2021-06-24 21:12:20.821][14][info][upstream] [source/common/upstream/cluster_manager_impl.cc:684] removing cluster tel-proxy-46655
[2021-06-24 21:12:20.821][14][info][upstream] [source/common/upstream/cds_api_impl.cc:91] cds: remove cluster 'tel-proxy-46655'
```

The interesting line is this - `time="2021-06-24 21:12:02.7497" level=error msg="tcp 127.0.0.1:59038 -> 127.0.0.1:8080: unhandled connection control message: tcp 127.0.0.1:59038 -> 127.0.0.1:8080, code CONNECT_REJECT, len 0" PID=1 THREAD=/sidecar lis=":0"`.  Why is it trying to connect to my local instance when I'm only establishing the intercept at this point?  
This creates a chicken and egg problem.  I need the intercept to succeed to get the env file so that I can start my local instance, but I need my local instance up and running before I can create the intercept.

**Comment by thallgren** [comment]
Could it be that there is some other service that is trying to access the service that you're intercepting?

What happens if you do have a service running on port 8080 on your local machine when you create the intercept? Does it receive a request? If so, what is it?

**Comment by foleymic** [comment]
If there is nothing running that is listening to port 8080 on my laptop, I get this error.  If I start anything that can receive the request on port 8080, e.g. - `docker run -d -p 8080:80 --name nginx nginx` then the intercept works.  Also, this is only happening if I am logged in to Ambassador (where I use the preview-url or http-match options).  This uses the proprietary agent, so I can't debug the agent code.  If I am  not logged into Ambassador, the OSS agent gets used and I do not see this issue, but then all requests are intercepted.

**Comment by thallgren** [comment]
Oh, sorry. Now I see what's happening. When you are using a preview-url (default when you're logged in), then the agent will do an initial check if the service supports http2. For some reason, that check takes very long to complete when there's no response at all at the other end. We'll need to investigate why that is.

**Comment by foleymic** [comment]
thanks @thallgren.  Is there something with my setup that can be contributing to this?  I'm just wondering why no one else has reported this issue.

**Comment by cyrus-mc** [comment]
I can confirm that I encountered this same error. 

For completeness the intercept command I am running is:

`telepresence intercept helloworld --port 3000 -u=false --http-match=auto -- yarn run start`

This resulted in the same error noted in the `traffic-agent` above. 

I have to start the local service first (`yarn run start`) before enabling the intercept. I also find that if I get into this error situation I can't recover without first removing traffic-agent (so restore the deployment to is original un-mutated state).

**Comment by msenmurugan** [comment]
getting the same error with v2.3.6

`telepresence: error: Intercept "c1aa0b2b-bab3-472b-a5d0-47e0153d37bb:auth-server-default": Failed to establish intercept: intercept in error state NO_AGENT: No agent found for "auth-server"`

2021/07/23 14:40:14 Patching synced Namespace 955c9055-41a0-11e9-9849-4201c0a8ff08
2021/07/23 14:40:17.0337 debug   connector/background-manager/intercept-port-forward : wait status: intercept id="c1aa0b2b-bab3-472b-a5d0-47e0153d37bb:auth-serve-default" is no longer WAITING; is now NO_AGENT (from pkg/client/connector/userd_trafficmgr/intercept.go:171)

telepresence list shows the state as WAITING

```
telepresence list
auth-server           : intercepted
    Intercept name         : auth-server-newtriage
    State                  : WAITING
    Workload kind          : Deployment
    Destination            : 127.0.0.1:8080
    Service Port Identifier: 8080
    Intercepting           : using mechanism="http" with args=["--match=auto"]
```

**Comment by efunk** [comment]
We resolved this with Telepresence 2.3.7 GA and Smart Agent 1.9.1.
